### ISSUE TITLE
Introduce an explicit Boolean type

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#17494]
So far, OpenSSL simply reuses `int` to represent Boolean values in the classical C manner:
`0` encodes `false`, and any nonzero value encodes `true`.
This introduces ambiguities that can cause quite some confusion and even errors:
* There are many cases where it is not clear whether, e.g., a function parameter or result is meant as an actual integer value or as a Boolean value.
* A rather often used convention is that a basically Boolean function can also return `-1` (or even further values) to indicate special (error) situations, and calls wrongly interpreting the result as a Boolean will go astray.

For such reasons, C++, Java, and most other languages distinguish (more or less clearly) between `int` and `bool`.
I propose doing (in the long run) the same within OpenSSL.

A simple way of achieving (most of) this is to declare the following:
```
#ifndef __cplusplus
typedef enum
{
    false = 0,
    true = 1
} bool; /*!< Boolean value */
#endif
```
This coexists seamlessly with the current state of OpenSSL,
which I can state from experience with its use also in https://github.com/siemens/libsecutils/blob/master/include/secutils/basic.h#L21.

**Comment by t8m** [comment]
You cannot do this for public interfaces because the bool, false, and true identifiers would frequently collide with other software. We would have to use OSSL_BOOL, OSSL_FALSE, OSSL_TRUE.

**Comment by t8m** [comment]
Also if you think the existing API calls should be converted to use this type, this would be 4.0 material.

**Comment by mattcaswell** [comment]
enums are not allowed in public interfaces. You could just "define" the true and false values.

**Comment by DDvO** [comment]
> enums are not allowed in public interfaces. You could just "`#define`" the `true` and `false` values.

Yes, and the subtype can alternatively be defined using `typedef`.

**Comment by DDvO** [comment]
> You cannot do this for public interfaces because the `bool,` `false,` and `true` identifiers would frequently collide with other software. We would have to use `OSSL_BOOL,` `OSSL_FALSE,` `OSSL_TRUE.`

Hmm, such `OSSL_` names would be pretty ugly IMHO and rather lengthy.

**Comment by DDvO** [comment]
> You cannot do this for public interfaces because the `bool,`... identifiers would frequently collide with other software.

In terms of having less chance of clashes and in order to have a replacement name that is equally short as `int` and would even avoid re-formatting, 
one might consider `boo`, but people won't like that :rofl:

**Comment by t8m** [comment]
Setting 4.0.0 milestone as IMO we do not want to introduce this without bigger API rehash thing.

**Comment by levitte** [comment]
> enums are not allowed in public interfaces. You could just "define" the `true` and `false` values.

I would like to revisit that rule, as it's too general.

**Comment by dlegaultbbry** [comment]
Most operating systems already have a bool type which is a define of the C99 standard _Bool type. Is there a reason you can't use stdbool.h and the defined true/false values that come with it and just define these for platforms which don't have this?

**Comment by paulidale** [comment]
I agree that we need to distinguish functions that return 0/1 for false & true from functions that return -1/0/1 or worse.  There have been a stream of fixes recently addressing exactly this kind of thing.

The problem is how.  Ugly, conflicting or both seems to apply to all solutions.

Defining _Bool if we're pre-C99 seems like it might work.  There will bit a pile of exceptional cases for compilers that have it already but claim to be pre-C99but this ugly would appear in one place.

**Comment by h-vetinari** [comment]
> Defining _Bool if we're pre-C99 seems like it might work.

Or dropping pre-C99** completely? It's been 22+ years - even for OpenSSL, that should be enough...?

** Yes, if there were a new baseline, it'd have to be C99-minus-all-the-stuff-that-became-optional-in-C11...

**Comment by levitte** [comment]
> Or dropping pre-C99\*\* completely? It's been 22+ years - even for OpenSSL, that should be enough...?

This has been argued already, but there are folks out there that are stuck with ANSI C.

**Comment by paulidale** [comment]
We officially use C89 not ANSI.  Not that there is much difference.
Plus we've got contamination with ANSI features.

**Comment by h-vetinari** [comment]
> This has been argued already

Is there a link to the discussion available?

> [...] but there are folks out there that are stuck with ANSI C.

Sure, but that number will never reach 0 until the extinction of the human race. Is there truly never a cut-off where the long tail becomes thin enough so that the cost of dropping it becomes less than the benefits of moving on?

**Comment by DDvO** [comment]
I also would be very glad if C89 / ANSI C support was dropped.
Retaining compatibility with that old stuff causes various complications.

How about setting up an independent issue for discussing and deciding on this?

**Comment by levitte** [comment]
> > [...] but there are folks out there that are stuck with ANSI C.
> 
> Sure, but that number will never reach 0 until the extinction of the human race. Is there truly never a cut-off where the long tail becomes thin enough so that the cost of dropping it becomes less than the benefits of moving on?

I really can't say much about C89...  I do know an operating system that is stuck with C95 on current architectures (C-much-more-modern is to appear on new architectures, though)

**Comment by GowthamKudupudi** [comment]
may be voyager is using pre c99, we don't want to loose it

**Comment by marka63** [comment]
> may be voyager is using pre c99, we don't want to loose it

Voyager is programmed in FORTRAN.  If it was programmed in C it would have been K&R C in 1977.

**Comment by t-j-h** [comment]
> Voyager is programmed in FORTRAN. If it was programmed in C it would have been K&R C in 1977.

Actually the on-board stuff was programmed in assembler for each of the different components.
The ground based system was in FORTRAN 5 then FORTRAN 77 and then C.
(according to various sources but there seems to be some controversy about parts of the history there).

And definitely the system long predates OpenSSL ...

**Comment by DDvO** [comment]
And for sure developing Voyager was considered rocket science 😉

**Comment by DDvO** [comment]
Now is the time to take a decision on this.

**Comment by GowthamKudupudi** [comment]
is it a issue because openssl's Makefile compiles openssl with -std=c89?

**Comment by DDvO** [comment]
Hi again @openssl/committers, any chance to get this into 4.0?

**Comment by t8m** [comment]
> Hi again [@openssl/committers](https://github.com/orgs/openssl/teams/committers), any chance to get this into 4.0?

I think it is too late. However we could decide we start using bool for new API calls.

**Comment by DDvO** [comment]
> > Hi again [@openssl/committers](https://github.com/orgs/openssl/teams/committers), any chance to get this into 4.0?

Thank you for quick response.
 
> I think it is too late.

What a pity.
I had asked already in October, but looks like this was missed.

> However we could decide we start using bool for new API calls.

Well, better than nothing.
This we could have done anytime, already just after this was brought up in early 2022.

**Comment by t8m** [comment]
> This we could have done anytime, already just after this was brought up in early 2022.

Not really, as it requires moving to C99 which we did not do earlier than during this year.

However we will need to decide this first. Also getting an advice on the topic from Communities & ТACs would be appropriate.

**Comment by fwh-dc** [comment]
Could this be proposed as a feature branch? 

That would allow us to work on it in the 4.x cycle and mature it for 5.0? Ofc it is not maintenance free, but if the first part of the work was just identifying and correcting functions in the public interface that should return or use boolean arguments. Then when 5.0 is approaching we could update internal source code to take advantage of it.

I would expect it to be quite maintainable. What do others think?

**Comment by DDvO** [comment]
I recently noticed that the C23 standard has introduced the type `bool` along with the keywords `true` and `false`.
